// @arcjet/node/index.js
import * as core from 'arcjet';
import core__default from 'arcjet';
export * from 'arcjet';
import { parseProxy, findIp } from '@arcjet/ip';
import { ArcjetHeaders } from '@arcjet/headers';
import { logLevel, isDevelopment, baseUrl, platform } from '@arcjet/env';
import { Logger } from '@arcjet/logger';
import { createClient } from '@arcjet/protocol/client.js';
import { createTransport } from '@arcjet/transport';
import { readBody } from '@arcjet/body';

// An object with getters that access the `process.env.SOMEVAR` values directly.
// This allows bundlers to replace the dot-notation access with string literals
// while still allowing dynamic access in runtime environments.
const env = {
    get FLY_APP_NAME() {
        return process.env.FLY_APP_NAME;
    },
    get VERCEL() {
        return process.env.VERCEL;
    },
    get RENDER() {
        return process.env.RENDER;
    },
    get MODE() {
        return process.env.MODE;
    },
    get NODE_ENV() {
        return process.env.NODE_ENV;
    },
    get ARCJET_ENV() {
        return process.env.ARCJET_ENV;
    },
    get ARCJET_LOG_LEVEL() {
        return process.env.ARCJET_LOG_LEVEL;
    },
    get ARCJET_BASE_URL() {
        return process.env.ARCJET_BASE_URL;
    },
    get FIREBASE_CONFIG() {
        return process.env.FIREBASE_CONFIG;
    },
};
let warnedForAutomaticBodyRead = false;
/**
 * Create a remote client.
 *
 * @param options
 *   Configuration (optional).
 * @returns
 *   Client.
 */
function createRemoteClient(options) {
    const url = options?.baseUrl ?? baseUrl(env);
    const timeout = options?.timeout ?? (isDevelopment(env) ? 1000 : 500);
    // Transport is the HTTP client that the client uses to make requests.
    const transport = createTransport(url);
    const sdkStack = "NODEJS";
    const sdkVersion = "1.4.0";
    return createClient({
        transport,
        baseUrl: url,
        timeout,
        sdkStack,
        sdkVersion,
    });
}
function cookiesToString(cookies) {
    if (typeof cookies === "undefined") {
        return "";
    }
    // This should never be the case with a Node.js cookie header, but we are safe
    if (Array.isArray(cookies)) {
        return cookies.join("; ");
    }
    return cookies;
}
/**
 * Create a new Node.js integration of Arcjet.
 *
 * > 👉 **Tip**:
 * > build your initial base client with as many rules as possible outside of a
 * > request handler;
 * > if you need more rules inside handlers later then you can call `withRule()`
 * > on that base client.
 *
 * @template Rules
 *   List of rules.
 * @template Characteristics
 *   Characteristics to track a user by.
 * @param options
 *   Configuration.
 * @returns
 *   Node.js integration of Arcjet.
 */
function arcjet(options) {
    const client = options.client ?? createRemoteClient();
    const log = options.log
        ? options.log
        : new Logger({
            level: logLevel(env),
        });
    const proxies = Array.isArray(options.proxies)
        ? options.proxies.map(parseProxy)
        : undefined;
    if (isDevelopment(env)) {
        log.warn("Arcjet will use 127.0.0.1 when missing public IP address in development mode");
    }
    function toArcjetRequest(request, props) {
        // We pull the cookies from the request before wrapping them in ArcjetHeaders
        const cookies = cookiesToString(request.headers?.cookie);
        // We construct an ArcjetHeaders to normalize over Headers
        const headers = new ArcjetHeaders(request.headers);
        const xArcjetIp = isDevelopment(env)
            ? headers.get("x-arcjet-ip")
            : undefined;
        let ip = xArcjetIp ||
            findIp({
                socket: request.socket,
                headers,
            }, { platform: platform(env), proxies });
        if (ip === "") {
            // If the `ip` is empty but we're in development mode, we default the IP
            // so the request doesn't fail.
            if (isDevelopment(env)) {
                ip = "127.0.0.1";
            }
            else {
                log.warn(`Client IP address is missing. If this is a dev environment set the ARCJET_ENV env var to "development"`);
            }
        }
        const method = request.method ?? "";
        const host = headers.get("host") ?? "";
        let path = "";
        let query = "";
        let protocol = "";
        if (typeof request.socket?.encrypted !== "undefined") {
            protocol = request.socket.encrypted ? "https:" : "http:";
        }
        else {
            protocol = "http:";
        }
        // Do some very simple validation, but also try/catch around URL parsing
        if (typeof request.url !== "undefined" &&
            request.url !== "" &&
            host !== "") {
            try {
                const url = new URL(request.url, `${protocol}//${host}`);
                path = url.pathname;
                query = url.search;
                protocol = url.protocol;
            }
            catch {
                // If the parsing above fails, just set the path as whatever url we
                // received.
                path = request.url ?? "";
                log.warn('Unable to parse URL. Using "%s" as `path`.', path);
            }
        }
        else {
            path = request.url ?? "";
        }
        return {
            ...props,
            ip,
            method,
            protocol,
            host,
            path,
            headers,
            cookies,
            query,
        };
    }
    function withClient(aj) {
        const client = {
            withRule(rule) {
                const client = aj.withRule(rule);
                return withClient(client);
            },
            async protect(request, props) {
                // Cast of `{}` because here we switch from `undefined` to `Properties`.
                const req = toArcjetRequest(request, props || {});
                const getBody = async () => {
                    // Read the stream if the body is not present.
                    if (request.body === null || request.body === undefined) {
                        let expectedLength;
                        // TODO: This shouldn't need to build headers again but the type
                        // for `req` above is overly relaxed
                        const headers = new ArcjetHeaders(request.headers);
                        const expectedLengthStr = headers.get("content-length");
                        if (typeof expectedLengthStr === "string") {
                            expectedLength = parseInt(expectedLengthStr, 10);
                        }
                        if (!warnedForAutomaticBodyRead) {
                            warnedForAutomaticBodyRead = true;
                            log.warn("Automatically reading the request body is deprecated; please pass an explicit `sensitiveInfoValue` field. See <https://docs.arcjet.com/upgrading/sdk-migration>.");
                        }
                        return readBody(request, { expectedLength });
                    }
                    // A package like `body-parser` was used to read the stream.
                    if (typeof request.body === "string") {
                        return request.body;
                    }
                    return JSON.stringify(request.body);
                };
                return aj.protect({ getBody }, req);
            },
        };
        return Object.freeze(client);
    }
    const aj = core__default({ ...options, client, log });
    return withClient(aj);
}

export { createRemoteClient, arcjet as default };
