// @arcjet/protocol/client.js
import { create } from '@bufbuild/protobuf';
import { createClient as createClient$1 } from '@connectrpc/connect';
import { DecideService, ReportRequestSchema, DecideRequestSchema } from './proto/decide/v1alpha1/decide_pb.js';
import { ArcjetStackToProtocol, ArcjetRuleToProtocol, ArcjetDecisionToProtocol, ArcjetDecisionFromProtocol } from './convert.js';
import 'typeid-js';

// TODO: Dedupe with `errorMessage` in core
function errorMessage(err) {
    if (err) {
        if (typeof err === "string") {
            return err;
        }
        if (typeof err === "object" &&
            "message" in err &&
            typeof err.message === "string") {
            return err.message;
        }
    }
    return "Unknown problem";
}
/**
 * Compute the timeout for a `Decide` request based on the configured rules.
 *
 * @internal Exported for testing only.
 * @param timeout
 *   Base timeout in milliseconds.
 * @param rules
 *   Rules that will be evaluated in this request.
 * @returns
 *   Adjusted timeout in milliseconds.
 */
function decideTimeout(timeout, rules) {
    let hasEmail = false;
    let hasPromptInjection = false;
    for (const rule of rules) {
        if (rule.type === "EMAIL") {
            hasEmail = true;
        }
        if (rule.type === "PROMPT_INJECTION_DETECTION") {
            hasPromptInjection = true;
        }
    }
    // If an email rule is configured, we double the timeout.
    // See https://github.com/arcjet/arcjet-js/issues/1697
    let result = hasEmail ? timeout * 2 : timeout;
    if (hasPromptInjection) {
        // We document the latency of this rule independently from other
        // `protect` calls, so we enforce a minimum timeout of 1 second.
        result = Math.max(result, 1_000);
    }
    return result;
}
function createClient(options) {
    const { transport, sdkVersion, baseUrl, timeout } = options;
    const sdkStack = ArcjetStackToProtocol(options.sdkStack);
    const client = createClient$1(DecideService, transport);
    return Object.freeze({
        async decide(context, details, rules) {
            const { log } = context;
            const protoRules = [];
            for (const rule of rules) {
                protoRules.push(ArcjetRuleToProtocol(rule));
            }
            const cleanDetails = {
                ip: details.ip,
                method: details.method,
                protocol: details.protocol,
                host: details.host,
                path: details.path,
                headers: Object.fromEntries(details.headers.entries()),
                cookies: details.cookies,
                query: details.query,
                extra: details.extra,
            };
            // Build the request object from the Protobuf generated class.
            const decideRequest = create(DecideRequestSchema, {
                sdkStack,
                sdkVersion,
                characteristics: context.characteristics,
                // `email` is an optional field but not allowed to be `undefined`.
                details: typeof details.email === "string"
                    ? { ...cleanDetails, email: details.email }
                    : cleanDetails,
                rules: protoRules,
            });
            log.debug("Decide request to %s", baseUrl);
            const response = await client.decide(decideRequest, {
                headers: { Authorization: `Bearer ${context.key}` },
                timeoutMs: decideTimeout(timeout, rules),
            });
            const decision = ArcjetDecisionFromProtocol(response.decision);
            log.debug({
                id: decision.id,
                fingerprint: context.fingerprint,
                path: details.path,
                runtime: context.runtime,
                ttl: decision.ttl,
                conclusion: decision.conclusion,
                reason: decision.reason,
                ruleResults: decision.results,
            }, "Decide response");
            return decision;
        },
        report(context, details, decision, rules) {
            const { log } = context;
            const cleanDetails = {
                ip: details.ip,
                method: details.method,
                protocol: details.protocol,
                host: details.host,
                path: details.path,
                headers: Object.fromEntries(details.headers.entries()),
                cookies: details.cookies,
                query: details.query,
                extra: details.extra,
            };
            // Build the request object from the Protobuf generated class.
            const reportRequest = create(ReportRequestSchema, {
                sdkStack,
                sdkVersion,
                characteristics: context.characteristics,
                // `email` is an optional field but not allowed to be `undefined`.
                details: typeof details.email === "string"
                    ? { ...cleanDetails, email: details.email }
                    : cleanDetails,
                decision: ArcjetDecisionToProtocol(decision),
                rules: rules.map(ArcjetRuleToProtocol),
            });
            log.debug("Report request to %s", baseUrl);
            // We use the promise API directly to avoid returning a promise from this
            // function so execution can't be paused with `await`
            const reportPromise = client
                .report(reportRequest, {
                headers: { Authorization: `Bearer ${context.key}` },
                // Rules don't execute during `Report` so we don't adjust the timeout
                // if an email rule is configured.
                timeoutMs: 2_000, // 2 seconds
            })
                .then((response) => {
                log.debug({
                    id: decision.id,
                    fingerprint: context.fingerprint,
                    path: details.path,
                    runtime: context.runtime,
                    ttl: decision.ttl,
                }, "Report response");
            })
                .catch((err) => {
                log.info("Encountered problem sending report: %s", errorMessage(err));
            });
            if (typeof context.waitUntil === "function") {
                context.waitUntil(reportPromise);
            }
        },
    });
}

export { createClient, decideTimeout };
